File: app/src/main/java/org/metplus/curriculum/security/SecurityConfig.java
package org.metplus.curriculum.security;

import org.metplus.curriculum.security.filters.ApplicationLoginFilter;
import org.metplus.curriculum.security.filters.ApplicationTokenAuthenticationFilter;
import org.metplus.curriculum.security.services.LocalTokenService;
import org.metplus.curriculum.security.services.TokenService;
import org.metplus.curriculum.security.useCases.UserTryToLogin;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.web.servlet.config.annotation.InterceptorRegistry;
import org.springframework.web.servlet.config.annotation.WebMvcConfigurerAdapter;
import org.springframework.web.servlet.handler.MappedInterceptor;

@Configuration
public class SecurityConfig extends WebMvcConfigurerAdapter {
    @Value("${backend.admin.username}")
    String username;
    @Value("${backend.admin.password}")
    String password;

    @Override
    public void addInterceptors(InterceptorRegistry registry) {
        registry.addInterceptor(applicationLoginFilter()).addPathPatterns("/**/authenticate");
        registry.addInterceptor(applicationTokenAuthenticationFilter());
    }

    private ApplicationLoginFilter applicationLoginFilter() {
        return new ApplicationLoginFilter(useCaseUserTryToLogin(), localTokenService());
    }

    private ApplicationTokenAuthenticationFilter applicationTokenAuthenticationFilter() {
        return new ApplicationTokenAuthenticationFilter(localTokenService());
    }

    private UserTryToLogin useCaseUserTryToLogin() {
        return new UserTryToLogin(username, password);
    }

    @Bean
    public TokenService localTokenService() {
        return new LocalTokenService();
    }
}

File: web/src/main/kotlin/org/metplus/cruncher/web/security/SecurityConfig.kt
package org.metplus.cruncher.web.security


import org.metplus.cruncher.web.security.filters.ApplicationLoginFilter
import org.metplus.cruncher.web.security.filters.ApplicationTokenAuthenticationFilter
import org.metplus.cruncher.web.security.services.LocalTokenService
import org.metplus.cruncher.web.security.services.TokenService
import org.metplus.cruncher.web.security.useCases.UserTryToLogin
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.web.servlet.config.annotation.InterceptorRegistry
import org.springframework.web.servlet.config.annotation.WebMvcConfigurer

@Configuration
open class SecurityConfig : WebMvcConfigurer {
    @Value("\${backend.admin.username}")
    internal var username: String? = null
    @Value("\${backend.admin.password}")
    internal var password: String? = null

    override fun addInterceptors(registry: InterceptorRegistry?) {
        registry!!.addInterceptor(applicationLoginFilter()).addPathPatterns("/api/v*/authenticate")
        registry.addInterceptor(applicationTokenAuthenticationFilter())
    }

    private fun applicationLoginFilter(): ApplicationLoginFilter {
        return ApplicationLoginFilter(useCaseUserTryToLogin(), localTokenService())
    }

    private fun applicationTokenAuthenticationFilter(): ApplicationTokenAuthenticationFilter {
        return ApplicationTokenAuthenticationFilter(localTokenService())
    }

    private fun useCaseUserTryToLogin(): UserTryToLogin {
        return UserTryToLogin(username, password)
    }

    @Bean
    open fun localTokenService(): TokenService {
        return LocalTokenService()
    }
}
